### PR TITLE
Add unix install script for smithy-cli

### PR DESCRIPTION
**Pull request by haydenbaker**
*Issue #, if available:*
N/A

*Description of changes:*
Adds install scripts to be used to install the smithy-cli distributable (and is included within it)

*Testing*
- Build the smithy-cli distributables via `gradlew :smithy-cli:runtime`, and navigate to the respective image for your machine (`smithy-cli/build/image/smithy-cli-<platfrm>/`)

For the shell script, you can:
- Run the install script via `sudo ./install`
- Optionally, you may configure a custom install and bin directory to use instead of the defaults.
- Run the `bats` tests, according to the README instructions (added in this PR)

For the batch script, you can:
- Generate the windows runtime, `./gradlew :smithy-cli:runtime`, and download it to a windows environment
- Run install.bat as administrator
- Or, you can do a mock installation according to the steps in the README

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-cli/build.gradle [hunk 1]
```diff
@@ -187,6 +187,18 @@ runtime {
     }
 }
 
+// Add finishing touches to the distributables, such as an install script
+tasks.runtime.doLast {
+    targetPlatforms.each { targetPlatform ->
+        copy {
+            from "config"
+            include targetPlatform.value.name.contains("windows") ? "install.bat" : "install"
```
Comment (mtdowling) [inline smithy-cli/build.gradle:195]: Where's the .bat file?

File: smithy-cli/config/install [hunk 2]
```diff
@@ -0,0 +1,135 @@
+#!/bin/sh
+
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (mtdowling) [inline smithy-cli/config/install:3]:
```suggestion
# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

Can we also use the shorter license header?

File: smithy-cli/config/install [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+create_bin_symlinks() {
+  echo "Setting up links..."
+  mkdir -p "$BIN_DIR"
+  ln -sf "$CURRENT_SMITHY_EXE" "$BIN_SMITHY_EXE"
+}
+
+quit() {
+	err_msg="$1"
```
Comment (mtdowling) [inline smithy-cli/config/install:120]: nit - indentation

File: smithy-cli/config/install [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                              for the Smithy CLI. By default, the directory
+                              used is: /usr/local/bin
+EOF
+}
+
+parse_commandline() {
+  while test $# -gt 0
+  do
+    key="$1"
+	case "$key" in
```
Comment (mtdowling) [inline smithy-cli/config/install:45]: nit - indentation

File: smithy-cli/config/install [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	exit 1
+}
+
+main() {
+  parse_commandline "$@"
+  set_global_vars
+  check_preexisting_install
+  create_install_dir
+  create_bin_symlinks
+  echo "You can now run: '$BIN_SMITHY_EXE --version' or '$EXE_NAME --version' if '$BIN_DIR' is already in your \$PATH"
```
Comment (mtdowling) [inline smithy-cli/config/install:131]: Can we detect if it's in ther path, and if not show a warning? And don't we need to invoke the warmup command too to add the CDS archive?

**Review by mtdowling — CHANGES_REQUESTED** [review]

File: smithy-cli/config/install [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  BIN_DIR=${PARSED_BIN_DIR:-/usr/local/bin}
+  UPGRADE=${PARSED_UPGRADE:-no}
+
+  EXE_NAME="smithy"
+  INSTALLER_DIR="$( cd "$( dirname "$0" )" >/dev/null 2>&1 && pwd )"
+  INSTALLER_EXE="$INSTALLER_DIR/bin/$EXE_NAME"
+  SMITHY_VERSION=$($INSTALLER_EXE --version)
+
+  INSTALL_DIR="$ROOT_INSTALL_DIR/$SMITHY_VERSION"
+#  INSTALL_DIR="$INSTALL_DIR"
```
Comment (haydenbaker) [inline smithy-cli/config/install:70]: Will remove this.

**Pull request merged**